crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs

```diff
@@ -249,16 +249,27 @@ impl Analyzer<'_, '_> {
                         if let Some(elem) = elem {
                             let elem_ty = ty
                                 .as_ref()
-                                .try_map(|ty| -> VResult<_> {
-                                    Ok(self
-                                        .get_element_from_iterator(span, Cow::Borrowed(&ty), idx)
-                                        .with_context(|| {
-                                            format!(
-                                                "tried to get the type of {}th element from iterator to declare vars with an array pattern",
-                                                idx
-                                            )
-                                        })?
-                                        .into_owned())
+                                .try_map(|ty| -> VResult {
+                                    let result = self.get_element_from_iterator(span, Cow::Borrowed(ty), idx).with_context(|| {
+                                        format!(
+                                            "tried to get the type of {}th element from iterator to declare vars with an array pattern",
+                                            idx
+                                        )
+                                    });
+
+                                    match result {
+                                        Ok(ty) => Ok(ty.into_owned()),
+                                        Err(err) => match err.actual() {
+                                            Error::TupleIndexError { .. } => {
+                                                if let RPat::Assign(prop) = elem {
+                                                    Ok(Type::any(prop.span, Default::default()))
```

Comment:
I'm not sure `any` is correct here

Reply:
This is wrong. We should use `default_ty` instead of `any`

Reply:
```ts
const [a, b = 1] = [1]; // ok
```

`b` should be number (generalized from 1) here, and this method accepts `default_ty` which can be used for it

Reply:
Oh sure..! Okay. I'll edit it and post it. thank you 🙇

Reply:
Oh, the first time I tried this, it failed.
When I took a picture of the current [L232 default](https://github.com/dudykr/stc/pull/202/files#diff-7b41dc3d40058bc0dceb3410349aab0fee895bf7ca944ee6a8c7766faea874dfL232), 
In the case of the current case, ‘None’ is displayed.

Maybe I got `default` wrong?

```
default: None
   WARN ((1:0)): Statement validation done. (time = 1.891458ms, kind: "perf", op: "validate (Stmt)"
    at crates/stc_ts_file_analyzer/src/analyzer/stmt/mod.rs:66
    in validate<RStmt>
    in validate<RModuleItem>
    in validate<RModule>

thread 'pass_tests__pass__exprs__array__destructuring_ts' panicked at 'Failed to validate.
error[TS2493]: DebugContext(
    context: tried to get the type of 1th element from iterator to declare vars with an array pattern
    context: tried to access property of a type to calculate element type
    context: tried to access property of an object ([1];, id_ctx = Var)
    Prop=Num(RNumber { span: Span { lo: BytePos(7), hi: BytePos(17), ctxt: #0 }, value: 1.0 })
    context: r-value context
    TupleIndexError {
        span: span {
            lo: BytePos(
                7,
            ),
            hi: BytePos(
                17,
            ),
            ctxt: #0,
        },
        len: 1,
        index: 1,
    },
)
 --> /tests/pass/exprs/array/destructuring.ts:1:7
  |
1 | const[a, b = a] = [1]; // ok
```

Reply:
Oh I think I found it above.
https://github.com/dudykr/stc/blob/e59631e58fb04578142befb6e5dbc7ef1d52f9db/crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs#L149

It looks like it could be done by making `right` the default_ty, I'll fix it once

Reply:
You should recover from https://github.com/dudykr/stc/blob/e59631e58fb04578142befb6e5dbc7ef1d52f9db/crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs#L253-L260

In the closure, you can check if `default` has a useful value

Reply:
But there is no `default` in the current situation (`None` printed in the comment above).
Perhaps the default_type is unknown, and the type must be inferred from the value of `Pat::Assign(assign_pat)`, `AssignPat::right`, am I wrong?

Thanks as always for your replies!

Reply:
You should use `default_elem_ty` below. I think you can change the order of statements.
Also, I think `default` should not be `None`. If it's `None`, I think it's a bug.

Reply:
https://github.com/dudykr/stc/pull/202#discussion_r1014600968
In the comment, I did not take the `default_elem_ty` you mentioned, but since `default` is already `None`, `default_elem_ty` will also be `None`.

Reply:
How do you debug in this regard?
Right now, I am printing a variable with `println`, but there are many stdouts that are output, so it is difficult to understand quickly..

Reply:
Oh, my bad. It should be computed from `RPat::Assign` as you said. `default` is for

```js
const [[a, b] = [1, 2]] = []
```

where the handler for `a` or `b` does not have access to `[1, 2]`

Reply:
I typically use `dbg!` macro and `print_backtrace()` to debug.
As I reduce tests to very small cases, those were enough.

Reply:
So now I'm trying to do something like this.
https://github.com/dudykr/stc/blob/e59631e58fb04578142befb6e5dbc7ef1d52f9db/crates/stc_ts_file_analyzer/src/analyzer/scope/vars.rs#L149-L162
It seems that the above is a way to infer the type of `RAssignPat::right`.

By matching this in `Result<Cow<Type>, Error>` made of `get_element_from_iterator` + `with_context`
If it is `TupleIndexError`, it tries to infer the type of `RAssignPat::right` and pass it.

how about it?

Reply:
I think it's correct, but I think we should unconditionally invoke `right.validate_with_args` and use the result only if required, so the RHS of an assignment pattern can be validated regardless of tuple index error.

Reply:
> use the result only if required, so the RHS of an assignment pattern can be validated regardless of tuple index error.

I didn't quite understand this, but I tried to fix it first.